File: ocaml/lib/day_04.ml
let rec scan_file channel =
  match input_line channel with
  | line -> line :: scan_file channel
  | exception End_of_file -> []

let read_lines filename =
  let ic = open_in filename in
  let lines = scan_file ic in
  close_in ic;
  List.rev lines

let numbers s =
  s |> String.split_on_char ' '
  |> List.filter (fun x -> String.length x > 0)
  |> List.map int_of_string

let rec count_repeats l1 l2 =
  match l1 with
  | [] -> 0
  | x :: xs -> (if List.mem x l2 then 1 else 0) + count_repeats xs l2

let rec pow a n = match n with 0 -> 1 | _ -> a * pow a (n - 1)

let line_score_part_one l =
  let parts = l |> String.split_on_char ':' in
  let number_sets = List.nth parts 1 |> String.split_on_char '|' in
  let winning_numbers = numbers (List.nth number_sets 0) in
  let card_numbers = numbers (List.nth number_sets 1) in
  let repeats = count_repeats winning_numbers card_numbers in
  match repeats with 0 -> 0 | x -> pow 2 (x - 1)

let rec total_score f lines =
  match lines with [] -> 0 | x :: xs -> f x + total_score f xs

let part_one filename =
  let lines = read_lines filename in
  total_score line_score_part_one lines

File: rs/src/day04.rs
fn get_numbers(s: String) -> Vec<i32> {
    s.split_whitespace()
        .map(|x| x.parse::<i32>().unwrap())
        .collect()
}

fn count_repeats(v1: Vec<i32>, v2: Vec<i32>) -> i32 {
    v1.iter()
        .fold(0, |acc, x| if v2.contains(&x) { acc + 1 } else { acc })
}

fn pow(a: i32, n: i32) -> i32 {
    match n {
        0 => 1,
        _ => a * pow(a, n - 1),
    }
}

fn line_score_part_one(line: String) -> i32 {
    let parts: Vec<String> = line.split(':').map(String::from).collect();
    let number_sets: Vec<Vec<i32>> = parts[1]
        .split('|')
        .map(String::from)
        .map(get_numbers)
        .collect();

    pow(
        2,
        count_repeats(number_sets[0].clone(), number_sets[1].clone()),
    )
}

pub fn part_one(filename: &str) -> i32 {
    let lines = std::fs::read_to_string(filename)
        .unwrap()
        .lines()
        .map(|x| String::from(x))
        .collect::<Vec<String>>();

    lines
        .iter()
        .fold(0, |acc, l| acc + line_score_part_one(l.to_owned()))
}

pub fn part_two(filename: &str) -> i32 {
    let lines = std::fs::read_to_string(filename)
        .unwrap()
        .lines()
        .map(|x| String::from(x))
        .collect::<Vec<String>>();
    let mut copies = vec![1; lines.len()];
    for i in 0..lines.len() {
        let parts: Vec<String> = lines[i].split(':').map(String::from).collect();
        let number_sets: Vec<Vec<i32>> = parts[1]
            .split('|')
            .map(String::from)
            .map(get_numbers)
            .collect();
        let repeats = count_repeats(number_sets[0].clone(), number_sets[1].clone());
        for j in 1..=repeats {
            copies[i + j as usize] += copies[i];
        }
    }
    copies.iter().sum()
}
